src/main/java/com/evch/rrm/CustomExecutorService.java

```diff
@@ -0,0 +1,306 @@
+package com.evch.rrm;
+
+import lombok.Getter;
+import lombok.Setter;
+
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class CustomExecutorService implements ExecutorService {
+    @Getter
+    @Setter
+    private long waitingWorkersTimeoutMs = 900_000L; // 15 minutes
+    @Getter
+    @Setter
+    private long waitingForNewTasksTimeoutMs = 20;
+    private final List<Thread> workers;
+    private final BlockingQueue<Runnable> tasks;
+    private final int maxCorePoolSize;
+    private final boolean useVirtualThreads;
+    private volatile boolean isShutdown = false;
+    private volatile boolean isShutdownNow = false;
+    private volatile boolean isAwaitTermination = false;
+    private final Lock lock = new ReentrantLock();
+
+    public CustomExecutorService(int corePoolSize, boolean useVirtualThreads) {
+        if (corePoolSize < 1) {
+            throw new IllegalArgumentException("The number of core pools must be greater than 0");
+        }
+        this.maxCorePoolSize = corePoolSize;
+        this.workers = new LinkedList<>();
+        this.tasks = new LinkedBlockingQueue<>();
+        this.useVirtualThreads = useVirtualThreads;
+
+        startThreadsExecutor();
+    }
+
+    public static void main(String[] args) {
+        CustomExecutorService ces = new CustomExecutorService(1, false);
+        ces.testPerformanceComparison();
+        ces.testConcurrentExecution();
+        ces.testShutdownBehavior();
+    }
+
+    private void startThreadsExecutor() {
+        Thread.ofVirtual().start(() -> {
+            while (!isTerminated() || isAwaitTermination) {
+                try {
+                    if (workers.size() < maxCorePoolSize && !isShutdownNow) {
```

Comment:
in reality there's no purpose in reusing virtual threads. thus, maxCorePoolSize does not make much sense for virtual threads. it's ok to use them as dispensibles.